server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -3897,6 +3898,63 @@ public void testFlushOnIdle() throws Exception {
         closeShards(shard);
     }
 
+    @TestLogging(reason = "testing traces of concurrent flushes", value = "org.elasticsearch.index.engine.Engine:TRACE")
+    public void testFlushOnIdleConcurrentFlushDoesNotWait() throws Exception {
+        final MockLogAppender mockLogAppender = new MockLogAppender();
+        try {
+            IndexShard shard = newStartedShard();
+            final int manyDocs = randomIntBetween(50, 100);
+            for (int i = 0; i < manyDocs; i++) {
+                indexDoc(shard, "_doc", Integer.toString(i));
+            }
+
+            mockLogAppender.start();
+            Loggers.addAppender(LogManager.getLogger(Engine.class), mockLogAppender);
+
+            shard.flushOnIdle(0); // flush happens in the background using the flush threadpool
+            assertFalse(shard.isActive());
+
+            // Wait for log message that flush acquired lock immediately
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "should see first flush getting lock immediately",
+                    Engine.class.getCanonicalName(),
+                    Level.TRACE,
+                    "acquired flush lock immediately"
+                )
+            );
+            assertBusy(mockLogAppender::assertAllExpectationsMatched);
+
+            // While the previous flushOnIdle request is happening, immediately issue a second flush request with waitIfOngoing=false
```

Comment:
I think we need to block the original flush in order to be sure this happens?

Reply:
Hi @henningandersen , indeed, that would be best. For the moment, I've indexed a large-enough amount to ensure the flush is still ongoing. Would you have a recommendation about how to inject a block of the original flush just for the test (rather than in reality)?

Reply:
After some tips from @fcofdez , I found a nice way to do it. Feel free to review :)